test/extensions/transport_sockets/alts/tsi_handshaker_test.cc

```diff
@@ -0,0 +1,161 @@
+  server_handshaker_.next(client_sent);
+  EXPECT_NE(nullptr, server_result);
+  EXPECT_EQ("SERVER_FINISHED", TestUtility::bufferToString(server_sent).substr(4));
+
+  client_callbacks_.expectDone(TSI_OK, client_sent, client_result);
+  client_handshaker_.next(server_sent);
+  EXPECT_NE(nullptr, client_result);
+  EXPECT_EQ("", TestUtility::bufferToString(client_sent));
+
+  tsi_peer client_peer;
+  EXPECT_EQ(TSI_OK, tsi_handshaker_result_extract_peer(client_result.get(), &client_peer));
+  EXPECT_EQ(1, client_peer.property_count);
+  EXPECT_STREQ("certificate_type", client_peer.properties[0].name);
+  absl::string_view client_certificate_type{client_peer.properties[0].value.data,
+                                            client_peer.properties[0].value.length};
+  EXPECT_EQ("FAKE", client_certificate_type);
+
+  tsi_peer server_peer;
+  EXPECT_EQ(TSI_OK, tsi_handshaker_result_extract_peer(server_result.get(), &server_peer));
+  EXPECT_EQ(1, server_peer.property_count);
+  EXPECT_STREQ("certificate_type", server_peer.properties[0].name);
+  absl::string_view server_certificate_type{server_peer.properties[0].value.data,
+                                            server_peer.properties[0].value.length};
+  EXPECT_EQ("FAKE", server_certificate_type);
+
+  tsi_peer_destruct(&client_peer);
+  tsi_peer_destruct(&server_peer);
+}
+
+TEST_F(TsiHandshakerTest, IncompleteData) {
+  InSequence s;
+
+  Buffer::OwnedImpl server_sent;
+  Buffer::OwnedImpl client_sent;
+
+  CHandshakerResultPtr client_result;
+  CHandshakerResultPtr server_result;
+
+  client_callbacks_.expectDone(TSI_OK, client_sent, client_result);
+  client_handshaker_.next(server_sent); // Intially server_sent is empty.
+  EXPECT_EQ(nullptr, client_result);
+  EXPECT_EQ("CLIENT_INIT", TestUtility::bufferToString(client_sent).substr(4));
+
+  client_sent.drain(3); // make data incomplete
+  server_callbacks_.expectDone(TSI_INCOMPLETE_DATA, server_sent, server_result);
+  server_handshaker_.next(client_sent);
+  EXPECT_EQ(nullptr, client_result);
+  EXPECT_EQ("", TestUtility::bufferToString(server_sent));
+}
+
+TEST_F(TsiHandshakerTest, DeferredDelete) {
+  InSequence s;
+
+  TsiHandshakerPtr handshaker{new TsiHandshaker({tsi_create_fake_handshaker(0)}, dispatcher_)};
+  handshaker->deferredDelete();
+  EXPECT_EQ(dispatcher_.to_delete_.back().get(), handshaker.get());
+  handshaker.release();
+}
+
+TEST_F(TsiHandshakerTest, DeleteOnDone) {
+  InSequence s;
+
+  TsiHandshakerPtr handshaker(new TsiHandshaker({tsi_create_fake_handshaker(1)}, dispatcher_));
+  handshaker->setHandshakerCallbacks(client_callbacks_);
+
+  Buffer::OwnedImpl empty;
+  std::function<void()> done;
+
+  EXPECT_CALL(dispatcher_, post(_)).WillOnce(SaveArg<0>(&done));
+
+  handshaker->next(empty);
+  handshaker->deferredDelete();
+
+  EXPECT_NE(dispatcher_.to_delete_.back().get(), handshaker.get());
+  EXPECT_CALL(client_callbacks_, onNextDone_(_)).Times(0);
```

Comment:
It took me a little while to figure out what these two lines are testing. Can you add some comments about why client_callbacks_.onNextDone() is not expected to be called. And why done() need to be called the next line?

Reply:
Done.